src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,932 @@
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Safe implementation of the slots tracker state.
+//!
+//! This module provides a single-thread-safe implementation of the slots tracker.
+//! All methods take `&mut self`, making the borrowing rules enforce single-threaded access.
+
+use std::num::NonZeroU32;
+
+use crate::SlotRange;
+use crate::slot_set::{CoverageRelation, SlotSet};
+
+/// Represents the version state of slot configuration.
+///
+/// This enum encapsulates the version logic and avoids exposing magic numbers.
+/// The discriminant itself IS the version value, making this exactly 32 bits.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Version {
+    /// Stable version values: 1..=u32::MAX
+    /// Each value represents a specific version counter.
+    Stable(NonZeroU32),
+    /// Unstable state marker
+    /// Slots are available but configuration is changing.
+    Unstable,
+}
+// Ensure that Version can be stored as AtomicU32 for fast atomic access.
+const _: () = assert!(std::mem::size_of::<Version>() == std::mem::size_of::<u32>());
+
+impl Default for Version {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Version {
+    /// Creates a new Version initialized to version 1 (stable).
+    const fn new() -> Self {
+        // SAFETY: value is not zero
+        Self::Stable(unsafe { NonZeroU32::new_unchecked(1) })
+    }
+
+    /// Increments a stable version using wrapping arithmetic.
+    ///
+    /// # Panics
+    ///
+    /// Panics in debug mode if called on an Unstable version.
+    fn increment(self) -> Self {
+        match self {
+            Self::Stable(v) => {
+                if v.get() < u32::MAX {
+                    // SAFETY: 1 < v + 1 <= u32::MAX, so not zero
+                    Self::Stable(unsafe { NonZeroU32::new_unchecked(v.get() + 1) })
+                } else {
+                    // Wrap around to 1 from u32::MAX
+                    // SAFETY: value is not zero
+                    Self::Stable(unsafe { NonZeroU32::new_unchecked(1) })
+                }
+            }
+            Self::Unstable => {
+                debug_assert!(false, "Cannot increment Unstable version");
+                Self::Unstable
+            }
+        }
+    }
+}
+
+/// Safe slots tracker implementation.
+///
+/// This structure encapsulates all slot tracking state and provides safe methods
+/// for manipulating the three slot sets and version counter.
+///
+/// All methods take `&mut self`, ensuring compile-time enforcement of exclusive access.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct SlotsTracker {
+    /// Local responsibility slots - owned by this Redis instance in the cluster topology.
+    local: SlotSet,
+    /// Fully available non-owned slots - locally available but not owned by this instance.
+    fully_available: SlotSet,
+    /// Partially available non-owned slots - partially available and not owned by this instance.
```

Comment:
clarify partially available meaning here?

It means that not all keys of the range slot are in this instance?